<a href="https://colab.research.google.com/github/NilavanIniyavan/GlobalAIHubMLCourse/blob/main/Homework_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Question 1:
Import Boston Dataset from sklearn dataset class.

In [6]:
from sklearn.datasets import load_boston

Question 2:
Explore and analyse raw data.

In [7]:
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.datasets import make_regression
X_m, y_m = make_regression(n_samples=100, n_features=2, noise=20,random_state=42)
X_m.shape

(100, 2)

In [8]:
model = LinearRegression().fit(X_m, y_m)
r_sq = model.score(X_m, y_m)
print('coefficient of determination:', r_sq)
print('intercept:', model.intercept_)
print('slope:', model.coef_)

coefficient of determination: 0.959604849170989
intercept: 0.4327161689219041
slope: [84.26785337 74.15664645]


In [9]:
Xm_test, ym_test = make_regression(n_samples=20, n_features=2, noise=20, random_state=42)

ym_pred = model.predict(Xm_test)
print('predicted response:', ym_pred, sep='\n')

predicted response:
[ -53.2618421   -81.47203332  182.21191983 -185.91510539  -48.19449499
  -30.58967254 -142.85219232  -89.69922222  -68.68904094 -149.21585472
  -41.44798528   90.09462956  -73.17883242 -174.86416876 -114.62008988
  -36.66153161   11.33844031  111.89967811   25.61613614  176.80558272]


In [10]:
print('Coefficient of determination (R2) of test:', model.score(Xm_test,ym_test))

Coefficient of determination (R2) of test: 0.9587826719061529


Question 3:
Do preprocessing for regression.

In [11]:
import pandas as pd

Xb,yb =load_boston(return_X_y=True)

df_boston = pd.DataFrame(Xb,columns = load_boston().feature_names)
df_boston.head()

CRIM    ZN  INDUS  CHAS    NOX  ...  RAD    TAX  PTRATIO       B  LSTAT
0  0.00632  18.0   2.31   0.0  0.538  ...  1.0  296.0     15.3  396.90   4.98
1  0.02731   0.0   7.07   0.0  0.469  ...  2.0  242.0     17.8  396.90   9.14
2  0.02729   0.0   7.07   0.0  0.469  ...  2.0  242.0     17.8  392.83   4.03
3  0.03237   0.0   2.18   0.0  0.458  ...  3.0  222.0     18.7  394.63   2.94
4  0.06905   0.0   2.18   0.0  0.458  ...  3.0  222.0     18.7  396.90   5.33

[5 rows x 13 columns]

In [12]:
df_boston.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
dtypes: float64(13)
memory usage: 51.5 KB


In [13]:
df_boston.describe()

CRIM          ZN       INDUS  ...     PTRATIO           B       LSTAT
count  506.000000  506.000000  506.000000  ...  506.000000  506.000000  506.000000
mean     3.613524   11.363636   11.136779  ...   18.455534  356.674032   12.653063
std      8.601545   23.322453    6.860353  ...    2.164946   91.294864    7.141062
min      0.006320    0.000000    0.460000  ...   12.600000    0.320000    1.730000
25%      0.082045    0.000000    5.190000  ...   17.400000  375.377500    6.950000
50%      0.256510    0.000000    9.690000  ...   19.050000  391.440000   11.360000
75%      3.677083   12.500000   18.100000  ...   20.200000  396.225000   16.955000
max     88.976200  100.000000   27.740000  ...   22.000000  396.900000   37.970000

[8 rows x 13 columns]

In [14]:
df_boston.isna().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
dtype: int64

Question 4:
Split your dataset into train and test test (0.7 for train and 0.3 for test).

In [15]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(Xb,yb,train_size=0.7, test_size=0.3, random_state=42)
X_train.shape

(354, 13)

Question 5:
Try Ridge and Lasso Regression models with at least 5 different alpha value for each.

In [21]:
from sklearn.linear_model import LinearRegression, Ridge, Lasso

In [22]:
ridge_model = Ridge(alpha = 0.01)
ridge_model.fit(X_train, y_train)
lasso_model = Lasso(alpha = 0.01)
lasso_model.fit(X_train, y_train)
print("Ridge Train: ", ridge_model.score(X_train, y_train))
print("Ridge Test: ", ridge_model.score(X_test, y_test))
print('*************************')
print("Lasso Train: ", lasso_model.score(X_train, y_train))
print("Lasso Test: ", lasso_model.score(X_test, y_test))

Ridge Train:  0.7434990281567367
Ridge Test:  0.7111351633163316
*************************
Lasso Train:  0.743066482217795
Lasso Test:  0.7088264519692221


In [24]:
ridge_model = Ridge(alpha = 0.02)
ridge_model.fit(X_train, y_train)
lasso_model = Lasso(alpha = 0.02)
lasso_model.fit(X_train, y_train)
print("Ridge Train: ", ridge_model.score(X_train, y_train))
print("Ridge Test: ", ridge_model.score(X_test, y_test))
print('*************************')
print("Lasso Train: ", lasso_model.score(X_train, y_train))
print("Lasso Test: ", lasso_model.score(X_test, y_test))

Ridge Train:  0.7434969073574276
Ridge Test:  0.7110441931030493
*************************
Lasso Train:  0.7417666684550652
Lasso Test:  0.7052390841975381


In [25]:
ridge_model = Ridge(alpha = 0.001)
ridge_model.fit(X_train, y_train)
lasso_model = Lasso(alpha = 0.001)
lasso_model.fit(X_train, y_train)
print("Ridge Train: ", ridge_model.score(X_train, y_train))
print("Ridge Test: ", ridge_model.score(X_test, y_test))
print('*************************')
print("Lasso Train: ", lasso_model.score(X_train, y_train))
print("Lasso Test: ", lasso_model.score(X_test, y_test))

Ridge Train:  0.7434997458244244
Ridge Test:  0.7112169289517238
*************************
Lasso Train:  0.7434954195159733
Lasso Test:  0.7110394345928663


In [26]:
ridge_model = Ridge(alpha = 0.002)
ridge_model.fit(X_train, y_train)
lasso_model = Lasso(alpha = 0.002)
lasso_model.fit(X_train, y_train)
print("Ridge Train: ", ridge_model.score(X_train, y_train))
print("Ridge Test: ", ridge_model.score(X_test, y_test))
print('*************************')
print("Lasso Train: ", lasso_model.score(X_train, y_train))
print("Lasso Test: ", lasso_model.score(X_test, y_test))

Ridge Train:  0.7434997237527576
Ridge Test:  0.7112078503361525
*************************
Lasso Train:  0.7434824204188089
Lasso Test:  0.7108409922939348


In [27]:
ridge_model = Ridge(alpha = 0.05)
ridge_model.fit(X_train, y_train)
lasso_model = Lasso(alpha = 0.05)
lasso_model.fit(X_train, y_train)
print("Ridge Train: ", ridge_model.score(X_train, y_train))
print("Ridge Test: ", ridge_model.score(X_test, y_test))
print('*************************')
print("Lasso Train: ", lasso_model.score(X_train, y_train))
print("Lasso Test: ", lasso_model.score(X_test, y_test))

Ridge Train:  0.7434829299372396
Ridge Test:  0.7107710601788526
*************************
Lasso Train:  0.735044871451358
Lasso Test:  0.6916678628681786


Question 6:
Evaluate the results of all models and choose the best performing model.

1.Ridge Test:  0.7111351633163316

  Lasso Test:  0.7088264519692221

2.Ridge Test:  0.7110441931030493

  Lasso Test:  0.7052390841975381  

3.Ridge Test:  0.7112169289517238

  Lasso Test:  0.7110394345928663

4.Ridge Test:  0.7112078503361525

  Lasso Test:  0.7108409922939348

5.Ridge Test:  0.7107710601788526

Lasso Test:  0.6916678628681786


From the above results we can conclude that the model 3 with alpha value 0.001 is the best performing model.